### PR TITLE
Project infrastrcture

### PR DESCRIPTION
**Pull request by unimonkiez**
I'm planning to do these:
* [x] CI to do
  * [x] Format check
  * [x] Eslint check
  * [x] Type check
  * [x] Run jest tests
* [x] CD to
  * [x] publish to NPM only if commit is tagged with `npm version patch/minor/major`.
  * [x] Publish js files with typing files on the same repo.

### CODE REVIEW

**Comment by unimonkiez** [comment]
Done, also checked the [published package](https://www.npmjs.com/package/form-mighty) and it downloads and has types just fine.
Make sure to add a secret to the project named `NPM_TOKEN` with the token I'll provide you elsewhere to make sure CD works.

File: .vscode/extensions.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
+  "recommendations": ["dbaeumer.vscode-eslint", "esbenp.prettier-vscode"]
```
Comment (omerman) [inline .vscode/extensions.json:2]: <3

File: .vscode/settings.json [hunk 2]
```diff
@@ -0,0 +1,14 @@
+{
```
Comment (omerman) [inline .vscode/settings.json:1]: <3

File: package.json [hunk 3]
```diff
@@ -25,7 +26,15 @@
   "scripts": {
     "start": "react-scripts start",
     "build": "react-scripts build",
+    "build:lib": "json -I -f tsconfig.json -e 'this.compilerOptions.noEmit=false' && tsc --project tsconfig.json --declaration",
     "test": "react-scripts test",
+    "test:all": "react-scripts test --watchAll=false --passWithNoTests",
```
Comment (omerman) [inline package.json:31]: please change to `test:ci`,
also please add "tsc &&" as prefix for this command,
there is an open issue for cra team, they basically ignore ts errors in .test files... so we need this prefix.

File: src/lib/index.ts [hunk 4]
```diff
@@ -0,0 +1,5 @@
+export { FormProvider } from "./FormProvider";
+export { FormMighty } from "./FormMighty";
+export { FormSubscriber } from "./FormSubscriber";
+export { useInitForm } from "./useInitForm";
+export type { FormState } from "./types";
```
Comment (omerman) [inline src/lib/index.ts:5]: After my chanes this should be `export * from './types'`

**Comment by omerman** [comment]
Overall looks amazing. Please address my comments and feel free to rebase & merge afterwards!

File: package.json (hunk 3 above)
Comment (unimonkiez) [reply package.json:31]: But we are running `yarn typecheck` beforehand anyway.

Comment (omerman) [reply package.json:31]: Understood, so nvm

File: src/lib/index.ts (hunk 4 above)
Comment (unimonkiez) [reply src/lib/index.ts:5]: Can't
image

Comment (omerman) [reply src/lib/index.ts]: Remove the type keyword and it should work

Comment (omerman) [reply src/lib/index.ts]: >> `export * from './types'`

**Pull request merged**